secadvisor/pkg/extend_gremlin.go

```diff
@@ -0,0 +1,131 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy ofthe License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specificlanguage governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package pkg
+
+import (
+	"bytes"
+	"encoding/json"
+	"strings"
+	"text/template"
+	"time"
+
+	cache "github.com/pmylund/go-cache"
+	"github.com/skydive-project/skydive/api/client"
+	"github.com/skydive-project/skydive/contrib/exporters/core"
+	"github.com/skydive-project/skydive/logging"
+	"github.com/spf13/viper"
+)
+
+// extendGremlin is used for transform extensions using gremlin expressions
+// Syntax supported is:
+//	VAR_NAME=<gremlin expression with a substitution strings>
+// Substitution strings are defined according to golang template usage using {{ and }}
+// For example:
+//	AA_Name=G.V().Has('RoutingTables.Src','{{.Network.A}}').Values('Host')
+// The substitution string refers to fields that exist in the data provided by the particular transform
+// If needed, be sure to put quotes around the substitution results
+// Use only single quotes in the gremlin expression
+// Need to define `Extend` field as a map[string]interface{} subfield of `transform`
+// Structure in yml file is as follows:
+//   transform:
+//     type: secadvisor
+//     secadvisor:
+//       extend:
+//         - VAR_NAME1=<gremlin expression with substitution strings>
+//         - VAR_NAME2=<gremlin expression with substitution strings>
+type extendGremlin struct {
+	gremlinClient    *client.GremlinQueryHelper
+	gremlinExprCache *cache.Cache
+	gremlinTemplates []*template.Template
+	newVars          []string
+	nTemplates       int
+}
+
+func NewExtendGremlin(cfg *viper.Viper) *extendGremlin {
+	gremlinClient := client.NewGremlinQueryHelper(core.CfgAuthOpts(cfg))
+	extendStrings := cfg.GetStringSlice(core.CfgRoot + "transform.secadvisor.extend")
+	var gremlinTemplates []*template.Template
+	var newVars []string
+	var nTemplates int
+	nTemplates = 0
+	gremlinTemplates = make([]*template.Template, len(extendStrings))
+	newVars = make([]string, len(extendStrings))
+	// parse the extension expressions and save them for future use
+	for _, ee := range extendStrings {
+		logging.GetLogger().Debugf("extend gremlin expression: %s", ee)
+		substrings := strings.SplitN(ee, "=", 2)
+		vv, err := template.New("extend_template").Parse(substrings[1])
+		if err != nil {
+			logging.GetLogger().Errorf("NewExtendGremlin: template error: %s", err)
+			continue
+		}
+		gremlinTemplates[nTemplates] = vv
+		newVars[nTemplates] = substrings[0]
+		nTemplates++
+	}
+	p := &extendGremlin{
```

Comment:
just return (no need to set variable `p`)

Reply:
OK